### PR TITLE
Dependency Injection

### PR DESCRIPTION
**Pull request by jhollingworth**
This is a revised plan after feedback and discussions with @dariocravero @taion

**Core Types == [Action Creators, Queries, Stores, State Sources]**

There is a new type called Application which will replace Marty's singleton. Each instance of Application will have its own dispatcher. When you instantiate the application you create instances of all types you need. Core types expect the application to passed into their constructor.

```js
class Application extends Marty.Application {
  constructor() {
    super()

    this.userQueries = new UserQueries(this);
    this.userActions = new UserActions(this);
    this.userStore = new UserStore(this, new Something());
  }
}

var app = new Application();

app.dispatcher.register(action => {
  console.log('Action dispatched', action)
});

app.userActions.deleteUser(123);
```

You will be able to access the application instance within core types via ``this.app``

```js
class UserStore extends Marty.Store {
  constructor(app, something) {
    super(app);

    this.something = something;
  }

  getUser(id) {
    return this.fetch({
      id: id,
      remotely() {
        return this.app.userQueries.getUser(id);
      }
    });
  }
}
```

To access the application within components the application instance will be passed down via react context's. To allow this the Marty applications have ``bindTo(Component)`` which wraps the component with a component that passes the the context down.

```js
var User = app.bindTo(require('./component/user'));

React.render(<User />, ...)

class User extends React.Component {
  deleteUser() {
    this.context.app.userActions.deleteUser(123)
  }
}

// Convenience method for { app: React.PropTypes.instanceOf(Marty.Application) }
User.contextTypes = Marty.contextTypes;
```

Testing becomes much easier now

```js
new UserStore({
    app: {
        userQueries: {
            getUser: sinon.stub().returns(...)
        }
    }
})

MartyTestUtils.mockApplication(Component, {
  userActions: {
    createUser: sinon.spy()
  }
})
```

All [state functions](http://martyjs.org/api/top-level-api/#state) will be moved onto the application instance. [dehyrdation](http://martyjs.org/api/top-level-api/#dehydrate) and [rehydration](http://martyjs.org/api/top-level-api/#rehydrate) will look at all stores that have been added to the application.

To support migration from v0.9 we will have ``Marty.isASingleton``. When ``Marty.isASingleton = true`` it will have the same behaviour as v0.9. When false (default)

* It will not maintain an internal registry (``Marty.registry`` will return undefined)
* ``Marty.register`` will throw an error saying you should use applications instead
* ``Marty.create*`` (e.g. ``Marty.createStore``) will return types instead of instances
* ``Marty.renderToString`` will behave differently

```js
// v0.9 || v0.10 && Marty.isASingleton == true

Marty.renderToString({
  type: require('./views/user'),
  context: Marty.createContext(),
  props: { bar: 'bar' },
  timeout: 1000
}).then(res => console.log(res.html));


// v0.10 && Marty.isASingleton == false

var app = new Marty.Application();
var User = app.bindTo(require('./views/user'));

Marty.renderToString(<User bar='bar' />, { timeout: 1000 })
     .then(res => console.log(res.html))
```

**Things we'd like to do in the future**

``Marty.Application`` has ``register(id, clazz)`` that allows to you decentralise registration if you like.

```js
var Application = require('marty').Application;

Application.register('userQueries', UserQueries);

var app = new Application();

app.userQueries.getUser()
```

**Motivations for this design**

The goals are to

* Make it easier to manage multiple instances of Marty easily in an idiomatic way
* Simplify how Marty applications are built with no magic to make it easier to understand whats happening

We are going to move away from singletons altogether. Testing is more difficult and if they want to start doing isomorphism then you have to re-write all your code. The only argument for them is simplicity but I think the small amount of added complexity brings significant more value.

Having to register all types up front is a contentious one.

Arguments against it:

* It requires to remember to add/remove the instance from the application class
* If you forget to remove it then it can lead to redundant code being required in.

Arguments for it:

* You only have to know about stores up front. However it will be confusing to have one strategy for registering stores and another for all other types
* Adding/removing stores/queries/action creators is not something that happens very often. I don't see the need to optimize for a situation that might happen once or twice a month
* You can do automatic registration using things like [bulkify](https://github.com/substack/bulkify) if you want. You could also split registration into multiple files (e.g. ./stores/index.js, ./queries/index.js) so you don't have one giant file
* This is how [alt](http://alt.js.org/) and [flummox](http://acdlite.github.io/flummox) do it and no one seems to be complaining

**Deprecations**

* ``Registry``
* ``Context``
* ``Marty.createContext``
* ``Marty.createInstance``
* ``Marty.register``
* ``Marty.registry``
* ``Marty.dispatcher``

**Tasks**

* [x] Add `Marty.isASingleton` to switch Marty's behaviour
  * [ ] Add documentation http://martyjs.org/guides/marty-is-a-singelton.html
* [ ] Work out how best to do registration

### CODE REVIEW

**Comment by jhollingworth** [comment]
@martyjs/core what are your thoughts on this?

**Comment by dariocravero** [comment]
[on gitter](https://gitter.im/martyjs/marty?at=552942031323d8f509c4d1a2)

**Comment by taion** [comment]
In no particular order:

* I really like the idea you proposed on Gitter to just inject the app via context and use it as the service locator: https://gitter.im/martyjs/marty?at=552954381323d8f509c4d2bb. This is going to be much easier to understand for most use cases than doing DI.

* Your example of alt-style registration alongside definition looks a lot like the more verbose case here: https://github.com/martyjs/marty/issues/261#issuecomment-89988298, which I agree is not totally ideal.

One longer comment:

On registry v. application, I thought about this some more. I agree that in most cases the user shouldn't have to learn 2 concepts. But in your registry example, there would actually be 3 concepts:

1. The registry itself
2. Creating an application with that registry and binding to it
3. Using the application context to resolve the Flux dependencies

I think the concept that can most economically be eliminated there is actually the application! In most cases, you're going to be binding the root component to a single application, so why not make `app.bindTo` spin up a new context? That way your syntax for setting things up can look very similar to alt's, but because you already have the `app.bindTo` method, you can be singleton-free "for free". The `bindTo` method (perhaps named differently) would take care of making new contexts.

In other words, kill the `Registry` concept, make the `Application` do all the work of the `Registry`, and spin up a new `ApplicationContext` every time you call `Application.bindTo`.

If that sounds reasonable, I also think it would improve developer ergonomics for the `Marty` root itself to serve as a default instance of `Application`. Just saves some typing, per what you mentioned in your comment on the issue linked above.

**Comment by sporto** [comment]
re alt strategy shown above

I don't get how this avoids a singleton, each time you do `var app = require('../app');` you will just get the same `app` instance for all requests.

**Comment by taion** [comment]
@sporto

The alt strategy does have a singleton, which is why alt recommends that you use something like the Fluxible approach of registering all your stores/&c. in the constructor.

However, alt doesn't have a `bindTo` method. If you make the syntax be something like `app.bindTo(Component)`, you can, in `bindTo`, generate a new e.g. `ApplicationContext` that holds non-singleton instances of the stores and other objects registered to `app.

**Comment by sporto** [comment]
Re conversation in gitter. I like the idea of trying to be idiomatic an using `context` as the place where to inject the marty instance. Some comments about this:

- I wouldn't like this to be called `app`, for me app is something larger, I think `flux` or `marty` works well.
- I would like the injected instances to start with downcase e.g. `this.context.flux.userActions` not `this.context.flux.UserActions` for me the downcase in JS is a clear convention that indicates an instance;

**Comment by sporto** [comment]
@taion I see what you mean, both approach are quite valid, but I am thinking that making your own instance of marty is more explicit and probably better for testing. As opposed to have .bindTo do this.

**Comment by taion** [comment]
@sporto I think this is the most straightforward way to have non-centralized registration while not having actual singletons. The alt pattern of doing centralized registration on the app itself doesn't work unless you go through some hoops to make the singleton stores not be a problem.

**Comment by sporto** [comment]
> We are going to deprecate .for(this) in favor of constructor injection. To inject a dependency into the type you use inject

This is nice, maybe it could be `use` instead of `inject`. I have the impression that when you say `inject` you expect a completely new instance of something. So if both UserQueries and UserAPI inject `Foo` they should get a different foo, but this is not what we want. Just my idea of `inject`.

> Should we automatically listen to stores if they are injected into a container?

Hmm, I like the explicitness but don't like unnecessary boilerplate. Either way.

 > Should you be explicit about instantiating all types in an apps registry? Or is it OK to be lazy?

Lazy, having to register everything seems unnecessary. I prefer to just have to register the minimum necessary i.e. stores

**Comment by dariocravero** [comment]
@sporto we're doing a hangout today at 5PM GMT ([tentative time](https://gitter.im/martyjs/marty?at=5529a03327e4ff0c43e26e94)) to discuss these points if you're on for it. [Here's the hangout link](https://plus.google.com/hangouts/_/uxtemple.com/marty?authuser=0) /cc @martyjs/core

**Comment by jhollingworth** [comment]
@martyjs/core I've updated the PR based on discussions. Let me know if this makes sense and if there is anything this approach will prevent us from doing

**Comment by jhollingworth** [comment]
Thanks for all your feedback on this :)

**Comment by taion** [comment]
The one thing I would add is that this gives us the possibility in the future to add a static `register` method to the `Application` class that would then allow defining registrations in-line (rather than in a centralized way), along the lines of the basic `alt` example.

@sporto You initially brought up the idea of non-centralized registrations. What do you think?

**Comment by jhollingworth** [comment]
@taion Can't find link to your babel example, this is roughly what you were suggesting right?

```js
class Application extend Marty.Application {
  constructor() {
    super()
    this.userActions = new UserActions(this);
  }
}

Application.register('userQueries', UserQueries);

var app = new Application();

app.userQueries.getUser()
```

**Comment by taion** [comment]
Yup, and the signature looks something like

```js
class Application {
  static register(id, clazz) {
    ...
  }
}
```

This is roughly analogous with the basic alt example: https://github.com/goatslacker/alt#what-does-it-look-like, or with the current `Marty.register`. This is just how you would do it. I don't know if if makes sense to have both. alt has both: https://github.com/goatslacker/alt#flux-instances.

BTW, some state sources are immutable, if I'm not mistaken. For example, I think most HTTP state sources don't actually have any state, so they're actually fine to be singletons. Not sure if you think this is an interesting special case or not.

**Comment by jhollingworth** [comment]
@taion I've added a note about registration :smile: 

Agree that state sources are likely to be immutable however I can imagine some people assigning variables to the instance and having weird bugs. Also, would like to have one way of doing things everywhere, makes it easier to understand imho

```js
class UserAPI extends Marty.HttpStateSource {
  getUser(id) {
    if (this.lastId == id) {
      ...
    }

    this.lastId = id;

    this.get('/user')
  }
}
```

**Comment by taion** [comment]
Sure - just making a note of it.

The only place where this is even relevant is whether you pass the `app` into the state source. Per most of the examples in the documentation, it looks like maybe the state source should not have direct access to anything else in the app (and instead depend on returning a future to the action creator or else maybe taking a callback).

Not a big deal either way.

**Comment by sporto** [comment]
> When you instantiate the application you create instances of all types you need. 

Will this be the only way? I don't like this, sorry. I understand the need to do this for stores, but for everything?

> You will be able to access the application instance within core types via this.app

I don't find `app` a good name, way to generic, I already use app for other things.

**Comment by sporto** [comment]
Apart from my two comments above, all these is looking really nice.

**Comment by sporto** [comment]
I still would love to have a way of registering / resolving dependencies from anywhere. This follows better the modular pattern established by NPM and React.

```js
this.marty.resolve(UserActions).deleteUser()
```

Maybe it could be a more generic word, that implies register or resolve in one.

```js
this.marty.get(UserActions).deleteUser()
```

Now, if we were to have this, then I wouldn't see the need to have both something like `marty.get(UserQueries)` and `new UserQueries(this)`. I would probably be better to have just one pattern: `marty.get(UserQueries)`

The one thing that is nice about the `new UserQueries(this)` is that you can pass extra params in the constructor, you could to this in the other case by: `marty.get(UserQueries, x, y)`, but then only the first time you call `get` the params will be passed.

Anyway, please consider this.

**Comment by taion** [comment]
@sporto 

The thing that @jhollingworth brought up during the hangout was that it would ideal to be consistent as much as possible. It would be awkward for stores to be treated differently from action creators and queries - harder to explain, and harder to learn.

It's also pretty uncommon to add new Flux components, and with explicit registrations, you can actually do something like `this.context.flux.UserQueries`, which is fairly cool.

**Comment by jhollingworth** [comment]
> I don't find app a good name, way to generic, I already use app for other things.

How about if we allow you to configure the variable name? ``app`` would just be a sensible default

```js
var flux = new FluxApplication({
  name: 'flux'
});


class Home extends React.Component {
  addUser() {
    this.context.flux.userActions.deleteUser()
  }
}
```

> Will this be the only way? I don't like this, sorry. I understand the need to do this for stores, but for everything?

For v0.10 this will be the recommended way. As @taion says this is to do with simplifying things and making it easier for people to understand whats going on (e.g. having a consistent registration strategy throughout the application).

That said I don't see this as the final solution, merely a simple base that allows us to do more advanced registration strategies in the future. e.g. you could achieve decentralised registration easily with a couple of lines of code

```js
var app = require('./app');

function register(id, type) {
   app[id] = new type(app);
   
   return type;
}

class UserActionCreators extends Marty.ActionCreators {
}

module.exports = register('userActions', UserActionCreators);
```

**Comment by dariocravero** [comment]
I think that [use](https://github.com/dariocravero/simple-marty/blob/master/marty/core/app.js#L23-L33) could be added as an option so that use cases like that of @sporto can be supported.

It could be simplified as follows:

```js
// marty-core/app.js

import useFn from './use'

class App {
  constructor({name='app', use=false}) {
    if (use) {
      this.instances = {}
      this.use = useFn
    }
  }
}

// marty-core/use.js
export default function use(Module, options = {}) {
    const name = getName(Module)
    let instance = this.instances[name]

    if (!instance) {
      instance = this.instances[name] = new Module({context: {app: this}, displayName: name, ...options})
    }

    return instance
  }
```

Thoughts?

**Comment by sporto** [comment]
> How about if we allow you to configure the variable name? app would just be a sensible default

Nice

> I think that use could be added as an option so that use cases like that of ...

I will very much like this.

> ... and we won't do any extra checking to see whether you defined userQueries and are trying to use(UserQueries).

This is ok, as long it is very clear where is the name coming from in `getName(Module)`. Just a static .id attr attached to the class I guess?

**Comment by jhollingworth** [comment]
I think I have a simple solution which re-uses a lot most of our existing code and will be an easy migration path: We keep ``Marty.register`` which registers types to ``Marty.registry`` (Same as v0.9 except it doesn't return an instance). You then pass the registry in when constructing the application. 

```js
module.exports = Marty.register(UserStore, 'userStore');

var app = new Marty.Application({
  registry: Marty.registry
})

app.userStore.getUser()
```

**Comment by dariocravero** [comment]
I'd say that the sooner we get rid of the registry the better, so I'd be inclined not to have it in v0.10 if possible.

**Comment by jhollingworth** [comment]
Whats wrong with the registry? Also, we can't remove registry in v0.10. Any breaking changes should be warnings for one version so we can remove it in v0.11

**Comment by taion** [comment]
I was thinking of migration path when I mentioned calling exporting the application class as `Marty` instead of `Marty.Application`. If you then have a static `register` method, then migration becomes very easy, because registration then doesn't change much at all from how it currently is.

**Comment by jhollingworth** [comment]
My concern there is developers might have written code that uses the registry. Should we just remove it entirely? Or give developers a chance to migrate slowly over

**Comment by dariocravero** [comment]
I agree with you that we should keep it in line with a v0.10, that's why I hinted it would go up a major version when we remove it. I don't find anything inherently wrong with it, at the end of the day, `use` is a simplified version of it :), my point with getting away from it is that it's slightly complex and it served a purpose when isomorphism wasn't free but that won't be the case anymore.

**Comment by taion** [comment]
@jhollingworth You mean with adding a static `Marty.register` method that is actually on the base `Application` class? I guess I'm not familiar with all the use cases of the registry API that this might break.

**Comment by jhollingworth** [comment]
@dariocravero I'd like to avoid moving to v1.0 until both the Marty API has stabilised (We're definitely not there yet) and [React v1.0](https://facebook.github.io/react/blog/2014/03/28/the-road-to-1.0.html).

@taion It's more the fact there was a public API there in the first place which people might be using in their application. Up to now I've taken the view that changes to APIs should happen gradually, allowing people to upgrade but get lots of warnings they can correct. Perhaps thats just more work than its worth though?

**Comment by taion** [comment]
I don't know. The problem is that the proposed changes come pretty close to a re-write (for users) of all the places where user code touches Marty, outside of the definitions of the objects themselves. It's a lot more drastic than changing `displayName` to `id`, or having to remember to call `Marty.register` if I'm using ES6 classes (which wasn't an option in previous versions anyway).

I agree that e.g. the `UserActionCreators.for(this)` syntax is not ideal, though. Using `this.context.app.UserActionCreators` seems better but is actually more verbose.

There's a third option, actually. You could use https://www.npmjs.com/package/continuation-local-storage and make the current `Marty.register` return a proxy for the continuation-local object. Then you just code as if you were using singletons everywhere. This is completely different from everything we've talked about, and it seems pretty magical, but it leads to fairly nice user ergonomics. Flask does this on the backend side in Python, and I'm a fan of that setup, but it's easier to use than it is to really understand.

**Comment by sporto** [comment]
> Up to now I've taken the view that changes to APIs should happen gradually, allowing people to upgrade but get lots of warnings they can correct. Perhaps thats just more work than its worth though?

I think this is more work than it's worth. Decent strategy will be to keep all the deprecated methods around and then provide a good deprecation explanation and links when people try to use them, but these method wouldn't do anything apart from alerting the user. I believe react-router uses a strategy like this.

**Comment by jhollingworth** [comment]
This PR is being replaced with [this one](https://github.com/martyjs/marty-lib/pull/1). I've finished off pretty much everything we've discussed
